priam/src/main/java/com/netflix/priam/PriamServer.java

```diff
@@ -111,25 +115,48 @@ public void scheduleService() throws Exception {
                     UpdateSecuritySettings.getTimer(instanceIdentity));
         }
 
-        // Set up cassandra tuning.
-        cassandraTunerService.scheduleService();
+        // Set up the background configuration dumping thread
+        scheduleTask(
+                scheduler,
+                PriamConfigurationPersister.class,
+                PriamConfigurationPersister.getTimer(config));
+
+        boolean shouldStartCassandra = false;
 
-        // Determine if we need to restore from backup else start cassandra.
-        if (restoreContext.isRestoreEnabled()) {
+        // Determine if we need to restore from backup.
+        if (restoreContext.isRestoreEnabled(config, instanceIdentity.getInstanceInfo())) {
             restoreContext.restore();
-        } else { // no restores needed
-            logger.info("No restore needed, task not scheduled");
-            if (!config.doesCassandraStartManually()) cassProcess.start(true); // Start cassandra.
-            else
-                logger.info(
-                        "config.doesCassandraStartManually() is set to True, hence Cassandra needs to be started manually ...");
+            // Start cassandra only if restore is successful.
+            shouldStartCassandra = true;
+        } else {
+            if (instanceIdentity.isReplace()
+                    && backupRestoreConfig.enableBypassCassandraStreaming()) {
```

Comment:
How does Priam determine whether Cassandra hasn't successfully bootstrapped? I'm looked for an existing check, but I didn't see one.

Reply:
TokenRetrieverUtils.inferTokenOwnerFromGossip is used to fetch the instance identity. That method should tell correctly if Cassandra had already bootstrapped successfully.